File: ios/Cleanup/Features/Archive/ViewModels/ArchiveViewModel.swift
import SwiftUI
import Photos

@MainActor
class ArchiveViewModel: ObservableObject {
    @Published var archivedItems: [ArchivedItem] = []
    @Published var quota: ArchiveQuota?
    @Published var isLoading = false
    @Published var showSignIn = false
    @Published var showArchiveUpgradePaywall = false
    @Published var showArchiveSuccess = false
    @Published var archiveSuccessCount = 0
    @Published var showRestoreSuccess = false
    @Published var errorMessage: String?
    @Published var itemsToArchive: [ArchivableItem] = []

    private let archiveService = ArchiveService.shared
    private let authService = ArchiveAuthService.shared
    private let entitlementManager = EntitlementManager.shared

    // MARK: - Archivable Item Wrapper

    enum ArchivableItem {
        case photo(PhotoAsset)
        case video(VideoAsset)
        case contact(ContactItem)
        case vault(VaultItem)

        var fileSize: Int64 {
            switch self {
            case .photo(let p): return p.fileSize
            case .video(let v): return v.fileSize
            case .contact: return 1024
            case .vault(let v): return v.fileSize
            }
        }
    }

    var totalArchiveSize: Int64 {
        itemsToArchive.reduce(0) { $0 + $1.fileSize }
    }

    var totalArchiveSizeFormatted: String {
        ByteCountFormatter.string(fromByteCount: totalArchiveSize, countStyle: .file)
    }

    var archiveItemCount: Int {
        itemsToArchive.count
    }

    // MARK: - Filter

    enum ArchiveFilter: String, CaseIterable {
        case all = "All"
        case photos = "Photos"
        case videos = "Videos"
        case contacts = "Contacts"
    }

    @Published var selectedFilter: ArchiveFilter = .all

    var filteredItems: [ArchivedItem] {
        switch selectedFilter {
        case .all: return archivedItems
        case .photos: return archivedItems.filter { $0.fileType == .photo }
        case .videos: return archivedItems.filter { $0.fileType == .video }
        case .contacts: return archivedItems.filter { $0.fileType == .contact }
        }
    }

    // MARK: - Load

    func loadArchivedItems() {
        Task {
            isLoading = true
            archivedItems = await archiveService.getArchivedItems()
            do {
                quota = try await archiveService.getQuota()
            } catch {
                quota = archiveService.getCachedQuota()
            }
            isLoading = false
        }
    }

    func syncWithServer() {
        Task {
            do {
                try await archiveService.syncItems()
                archivedItems = await archiveService.getArchivedItems()
                quota = try await archiveService.getQuota()
            } catch {
                errorMessage = error.localizedDescription
            }
        }
    }

    // MARK: - Archive Actions

    func archiveSelectedPhotos(_ assets: [PhotoAsset]) {
        itemsToArchive = assets.map { .photo($0) }
        checkAuthAndArchive()
    }

    func archiveSelectedVideos(_ assets: [VideoAsset]) {
        itemsToArchive = assets.map { .video($0) }
        checkAuthAndArchive()
    }

    func archiveSelectedContacts(_ contacts: [ContactItem]) {
        itemsToArchive = contacts.map { .contact($0) }
        checkAuthAndArchive()
    }

    func archiveVaultItems(_ items: [VaultItem]) {
        itemsToArchive = items.map { .vault($0) }
        checkAuthAndArchive()
    }

    private func checkAuthAndArchive() {
        Task {
            let isReady = await authService.isFullySetUp
            if isReady {
                confirmArchive()
            } else {
                showSignIn = true
            }
        }
    }

    func onSignInComplete() {
        showSignIn = false
        confirmArchive()
    }

    func confirmArchive() {
        Task {
            isLoading = true
            errorMessage = nil

            // Ensure backend registration is complete and subscription is synced
            do {
                try await authService.ensureRegistered()
                try await authService.syncSubscription()
            } catch {
                errorMessage = "Failed to connect to archive service: \(error.localizedDescription)"
                isLoading = false
                return
            }

            // Pre-flight quota check using local subscription tier
            let quotaLimit: Int64
            if let tier = entitlementManager.currentArchiveTier {
                quotaLimit = tier.storageLimitBytes
            } else {
                quotaLimit = 0
            }

            var usedBytes: Int64 = 0
            do {
                let freshQuota = try await archiveService.getQuota()
                usedBytes = freshQuota.usedBytes
            } catch {
                if let cached = archiveService.getCachedQuota() {
                    usedBytes = cached.usedBytes
                }
            }

            let remainingBytes = max(0, quotaLimit - usedBytes)
            if quotaLimit > 0 && remainingBytes < totalArchiveSize {
                showArchiveUpgradePaywall = true
                isLoading = false
                return
            }

            // Archive all items WITHOUT deleting originals
            var archivedAssets: [PHAsset] = []
            var archiveSucceeded = false
            for item in itemsToArchive {
                do {
                    switch item {
                    case .photo(let asset):
                        _ = try await archiveService.archiveAsset(
                            asset.asset, tier: .instant, deleteAfterArchive: false
                        )
                        archivedAssets.append(asset.asset)
                    case .video(let asset):
                        _ = try await archiveService.archiveAsset(
                            asset.asset, tier: .instant, deleteAfterArchive: false
                        )
                        archivedAssets.append(asset.asset)
                    case .contact(let contact):
                        _ = try await archiveService.archiveContacts(
                            [contact], tier: .instant
                        )
                    case .vault(let vaultItem):
                        _ = try await archiveService.archiveVaultItem(
                            vaultItem, tier: .instant
                        )
                    }
                    archiveSucceeded = true
                } catch {
                    errorMessage = "Archive failed: \(error.localizedDescription)"
                    break
                }
            }

            // Batch delete originals with a single permission prompt AFTER all archives succeed
            if archiveSucceeded && !archivedAssets.isEmpty {
                do {
                    try await PHPhotoLibrary.shared().performChanges {
                        PHAssetChangeRequest.deleteAssets(archivedAssets as NSFastEnumeration)
                    }
                } catch {
                    // Deletion denied or failed — archive still succeeded, just don't delete originals
                }
            }

            isLoading = false

            if archiveSucceeded && errorMessage == nil {
                let count = itemsToArchive.count
                itemsToArchive.removeAll()
                archiveSuccessCount = count
                showArchiveSuccess = true
                loadArchivedItems()
            }
        }
    }

    // MARK: - Retrieval

    func initiateRetrieval(for item: ArchivedItem) {
        Task {
            do {
                let response = try await archiveService.initiateRetrieval(itemId: item.id)
                if response.status == "available", let urlString = response.downloadURL,
                   let url = URL(string: urlString) {
                    try await archiveService.downloadAndRestore(itemId: item.id, downloadURL: url)
                    // Remove from archive after successful restore
                    try? await archiveService.deleteArchivedItem(itemId: item.id)
                    archivedItems.removeAll { $0.id == item.id }
                    showRestoreSuccess = true
                }
                refreshQuota()
            } catch {
                errorMessage = error.localizedDescription
            }
        }
    }

    func checkRetrievalStatus(for item: ArchivedItem) {
        Task {
            do {
                let response = try await archiveService.checkRetrievalStatus(itemId: item.id)
                if response.status == "available", let urlString = response.downloadURL,
                   let url = URL(string: urlString) {
                    try await archiveService.downloadAndRestore(itemId: item.id, downloadURL: url)
                    try? await archiveService.deleteArchivedItem(itemId: item.id)
                    archivedItems.removeAll { $0.id == item.id }
                    showRestoreSuccess = true
                }
                refreshQuota()
            } catch {
                errorMessage = error.localizedDescription
            }
        }
    }

    // MARK: - Delete

    func deleteArchivedItem(_ item: ArchivedItem) {
        Task {
            do {
                try await archiveService.deleteArchivedItem(itemId: item.id)
                archivedItems.removeAll { $0.id == item.id }
                refreshQuota()
            } catch {
                errorMessage = error.localizedDescription
            }
        }
    }

    private func refreshQuota() {
        Task {
            do {
                quota = try await archiveService.getQuota()
            } catch {
                quota = archiveService.getCachedQuota()
            }
        }
    }
}

File: android/app/src/main/java/com/kreativekoala/cleanup/ui/archive/ArchiveViewModel.kt
package com.kreativekoala.cleanup.ui.archive

import android.net.Uri
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.kreativekoala.cleanup.billing.EntitlementManager
import com.kreativekoala.cleanup.data.model.ArchiveQuota
import com.kreativekoala.cleanup.data.model.ArchivedItem
import com.kreativekoala.cleanup.domain.service.ArchiveAuthService
import com.kreativekoala.cleanup.domain.service.ArchiveService
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * Port of iOS ArchiveViewModel.
 */
@HiltViewModel
class ArchiveViewModel @Inject constructor(
    private val archiveService: ArchiveService,
    private val authService: ArchiveAuthService,
    private val entitlementManager: EntitlementManager
) : ViewModel() {

    data class UiState(
        val quota: ArchiveQuota? = null,
        val isLoading: Boolean = false,
        val showSignIn: Boolean = false,
        val showUpgradePaywall: Boolean = false,
        val showArchiveSuccess: Boolean = false,
        val archiveSuccessCount: Int = 0,
        val showRestoreSuccess: Boolean = false,
        val errorMessage: String? = null,
        val selectedFilter: ArchiveFilter = ArchiveFilter.ALL,
        val selectedItem: ArchivedItem? = null,
        val showDeleteConfirmation: Boolean = false
    )

    enum class ArchiveFilter(val displayName: String) {
        ALL("All"),
        PHOTOS("Photos"),
        VIDEOS("Videos"),
        CONTACTS("Contacts")
    }

    private val _uiState = MutableStateFlow(UiState())
    val uiState: StateFlow<UiState> = _uiState.asStateFlow()

    val items: StateFlow<List<ArchivedItem>> = archiveService.getAllItems()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    val isAuthenticated: StateFlow<Boolean> = authService.isAuthenticated
    val isFullySetUp: StateFlow<Boolean> = authService.isFullySetUp

    init {
        loadData()
    }

    fun filteredItems(allItems: List<ArchivedItem>): List<ArchivedItem> {
        return when (_uiState.value.selectedFilter) {
            ArchiveFilter.ALL -> allItems
            ArchiveFilter.PHOTOS -> allItems.filter { it.fileType == "PHOTO" }
            ArchiveFilter.VIDEOS -> allItems.filter { it.fileType == "VIDEO" }
            ArchiveFilter.CONTACTS -> allItems.filter { it.fileType == "CONTACT" }
        }
    }

    // MARK: - Load

    fun loadData() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true) }
            try {
                val quota = archiveService.getQuota()
                _uiState.update { it.copy(quota = quota, isLoading = false) }
            } catch (_: Exception) {
                _uiState.update { it.copy(isLoading = false) }
            }
        }
    }

    fun syncWithServer() {
        viewModelScope.launch {
            try {
                archiveService.syncItems()
                val quota = archiveService.getQuota()
                _uiState.update { it.copy(quota = quota) }
            } catch (e: Exception) {
                _uiState.update { it.copy(errorMessage = e.localizedMessage) }
            }
        }
    }

    // MARK: - Archive

    fun archiveMedia(
        uri: Uri,
        fileName: String,
        fileType: String,
        fileSize: Long
    ) {
        viewModelScope.launch {
            if (!authService.isFullySetUp.value) {
                _uiState.update { it.copy(showSignIn = true) }
                return@launch
            }

            _uiState.update { it.copy(isLoading = true, errorMessage = null) }

            try {
                authService.ensureRegistered()
                archiveService.archiveMedia(
                    uri = uri,
                    fileName = fileName,
                    fileType = fileType,
                    fileSize = fileSize,
                    deleteAfterArchive = true
                )

                _uiState.update {
                    it.copy(
                        isLoading = false,
                        showArchiveSuccess = true,
                        archiveSuccessCount = 1
                    )
                }
                refreshQuota()
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(isLoading = false, errorMessage = "Archive failed: ${e.localizedMessage}")
                }
            }
        }
    }

    fun onSignInComplete() {
        _uiState.update { it.copy(showSignIn = false) }
    }

    fun signIn(googleIdToken: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true) }
            try {
                authService.signIn(googleIdToken)
                _uiState.update { it.copy(isLoading = false, showSignIn = false) }
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(isLoading = false, errorMessage = "Sign in failed: ${e.localizedMessage}")
                }
            }
        }
    }

    fun signOut() {
        authService.signOut()
    }

    // MARK: - Retrieval

    fun initiateRetrieval(item: ArchivedItem) {
        viewModelScope.launch {
            try {
                val response = archiveService.initiateRetrieval(item.id)
                if (response.status == "available" && response.downloadURL != null) {
                    archiveService.downloadAndRestore(item.id, response.downloadURL)
                    archiveService.deleteArchivedItem(item.id)
                    _uiState.update {
                        it.copy(showRestoreSuccess = true, selectedItem = null)
                    }
                }
                refreshQuota()
            } catch (e: Exception) {
                _uiState.update { it.copy(errorMessage = e.localizedMessage) }
            }
        }
    }

    // MARK: - Delete

    fun selectItem(item: ArchivedItem) {
        _uiState.update { it.copy(selectedItem = item) }
    }

    fun clearSelectedItem() {
        _uiState.update { it.copy(selectedItem = null) }
    }

    fun showDeleteConfirmation() {
        _uiState.update { it.copy(showDeleteConfirmation = true) }
    }

    fun dismissDeleteConfirmation() {
        _uiState.update { it.copy(showDeleteConfirmation = false) }
    }

    fun deleteSelectedItem() {
        val item = _uiState.value.selectedItem ?: return
        viewModelScope.launch {
            try {
                archiveService.deleteArchivedItem(item.id)
                _uiState.update {
                    it.copy(selectedItem = null, showDeleteConfirmation = false)
                }
                refreshQuota()
            } catch (e: Exception) {
                _uiState.update { it.copy(errorMessage = e.localizedMessage) }
            }
        }
    }

    // MARK: - Filter

    fun setFilter(filter: ArchiveFilter) {
        _uiState.update { it.copy(selectedFilter = filter) }
    }

    // MARK: - Dismiss

    fun dismissSuccess() {
        _uiState.update { it.copy(showArchiveSuccess = false, showRestoreSuccess = false) }
    }

    fun clearError() {
        _uiState.update { it.copy(errorMessage = null) }
    }

    fun showUpgradePaywall() {
        _uiState.update { it.copy(showUpgradePaywall = true) }
    }

    fun dismissUpgradePaywall() {
        _uiState.update { it.copy(showUpgradePaywall = false) }
    }

    private fun refreshQuota() {
        viewModelScope.launch {
            try {
                val quota = archiveService.getQuota()
                _uiState.update { it.copy(quota = quota) }
            } catch (_: Exception) {}
        }
    }
}
